Initiating PR for new Azure AD Recommendations API

hafowler opened the pull request.

FaithOmbongi commented on api-reference/beta/api/recommendation-list.md:
```diff
@@ -0,0 +1,132 @@
+---
```
Change file name to directory-list-recommendation.md

FaithOmbongi commented on api-reference/beta/api/recommendation-list.md:
```diff
@@ -0,0 +1,132 @@
+doc_type: apiPageType
+---
+
+# List recommendationBases
```
fix
```suggestion
# List recommendations
```

FaithOmbongi commented on api-reference/beta/api/recommendation-list.md:
```diff
@@ -0,0 +1,132 @@
+|Delegated (work or school account)|DirectoryRecommendations.Read.All, DirectoryRecommendations.ReadWrite.All|
+|Delegated (personal Microsoft account)|Not supported.|
+|Application|DirectoryRecommendations.Read.All, DirectoryRecommendations.ReadWrite.All|
+
```
GLOBAL: Please list all RBAC roles that a user/app must have to call this API. For the intro line, see this sample doc https://learn.microsoft.com/en-us/graph/api/identitygovernance-lifecycleworkflowscontainer-list-workflowtemplates?view=graph-rest-beta&tabs=http#permissions

FaithOmbongi commented on api-reference/beta/api/recommendation-list.md:
````diff
@@ -0,0 +1,132 @@
+```
+
+## Optional query parameters
+This method supports some of the OData query parameters to help customize the response. For general information, see [OData query parameters](/graph/query-parameters).
````
GLOBAL: All GET docs should explicitly the query parameters that this API supports.

FaithOmbongi commented on api-reference/beta/api/recommendation-list.md:
````diff
@@ -0,0 +1,132 @@
+  ]        
+}
+```
+
````
Suggestion: Add an example 2 that uses `$select` and `$filter` query parameters.

FaithOmbongi commented on api-reference/beta/api/recommendation-get.md:
````diff
@@ -0,0 +1,283 @@
+```
+
+## Optional query parameters
+This method supports some of the OData query parameters to help customize the response. For general information, see [OData query parameters](/graph/query-parameters).
````
$select, $expand?

FaithOmbongi replied:
@hafowler - bump

FaithOmbongi commented on api-reference/beta/api/recommendation-get.md:
```diff
@@ -0,0 +1,283 @@
+
+## Examples
+
+### Example 1: Get a recommendation object without its associated resources by the recommendationId
```
```suggestion
### Example 1: Get a recommendation object
```

FaithOmbongi commented on api-reference/beta/api/recommendation-get.md:
````diff
@@ -0,0 +1,283 @@
+  ]
+}
+```
+### Example 2: Get a recommendation object without its associated resources by the recommendationType
````
This is correct but in the wrong place. It should be documented in the LIST doc. You can move it as Example 2 in the LIST Recommendations doc but I also suggest tweaking the request to show use of $select as well ($select is the minimum query parameter that should be shown/supported by all APIs/API docs).

Considering example 3 above, I recommend you move this to the List doc, tweak to show use of $select, $filter, and $expand. Then Example 3 can show use of $select and $expand and remain in this doc as Example 2.

FaithOmbongi commented on api-reference/beta/api/recommendation-get.md:
````diff
@@ -0,0 +1,283 @@
+}
+```
+
+### Example 3: Get a recommendation object with its associated impacted resources
````
Also in the wrong doc. Remember the GET doc is for /recommendations/{id} basic syntaxes.

I recommend you change this URL to `https://graph.microsoft.com/beta/directory/recommendations/0cb31920-84b9-471f-a6fb-468c1a847088_Microsoft.Identity.IAM.Insights.TurnOffPerUserMFA/?$select=?,?,?$expand=impactedResources` (replace ? with the properties)

Then you can keep the new request as Example 2 in this doc.

Please make sure you run these requests in Graph Explorer.

FaithOmbongi commented on api-reference/beta/api/recommendation-postpone.md:
```diff
@@ -0,0 +1,140 @@
+## Request body
+In the request body, supply a JSON representation of the parameters.
+
+The following table shows the parameters that can be used with this action.
```
```suggestion
The following table shows the parameters that are required with this action.
```

FaithOmbongi replied:
@hafowler - bump

FaithOmbongi commented on api-reference/beta/api/recommendation-postpone.md:
```diff
@@ -0,0 +1,140 @@
+
+|Parameter|Type|Description|
+|:---|:---|:---|
+|postponeUntilDateTime|DateTimeOffset|**TODO: Add Description**|
```
Update description

FaithOmbongi replied:
@hafowler - bump

FaithOmbongi commented on api-reference/beta/api/recommendation-dismiss.md:
```diff
@@ -0,0 +1,140 @@
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Dismiss a [recommendation](../resources/recommendation.md) object that you consider to be inapplicable to your directory and update its **status** to `dismissed` .
```
```suggestion
Dismiss a [recommendation](../resources/recommendation.md) object that you consider to be inapplicable to your tenant and update its **status** to `dismissed` .
```

FaithOmbongi commented on api-reference/beta/api/recommendation-dismiss.md:
```diff
@@ -0,0 +1,140 @@
+## Request body
+In the request body, supply a JSON representation of the parameters.
+
+The following table shows the parameters that can be used with this action.
+
+|Parameter|Type|Description|
+|:---|:---|:---|
+|dismissReason|String|Explanation for why this recommendation is not applicable to your directory.|
```
```suggestion
The following table shows the parameters that are required with this action.

|Parameter|Type|Description|
|:---|:---|:---|
|dismissReason|String|Explanation for why this recommendation is not applicable to your tenant.|
```

FaithOmbongi commented on api-reference/beta/resources/impactedresource.md:
```diff
@@ -0,0 +1,82 @@
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Represents an Azure AD resource in your tenant that's associated with an Azure AD [recommendation](../resources/recommendation.md).
```
```suggestion
Represents an Azure AD resource in your tenant that's associated with an Azure AD [recommendation](../resources/recommendation.md).

Inherits from [entity](entity.md).
```

FaithOmbongi replied:
@hafowler - bump

FaithOmbongi commented on api-reference/beta/resources/impactedresource.md:
```diff
@@ -0,0 +1,82 @@
+## Properties
+|Property|Type|Description|
+|:---|:---|:---|
+|addedDateTime|DateTimeOffset|The date and time when the [impactedResource](../resources/impactedresource.md) object was initially associated with a [recommendation](../resources/recommendation.md).|
```
nit - because this is very specific to a particular recommendation. The impactedResoruce can be part of multiple recommendations but for the API, you get it only in the context of a particular recommendation. Please correct me if I'm wrong.
```suggestion
|addedDateTime|DateTimeOffset|The date and time when the [impactedResource](../resources/impactedresource.md) object was initially associated with the [recommendation](../resources/recommendation.md).|
```

FaithOmbongi replied:
@hafowler - bump

FaithOmbongi commented on api-reference/beta/resources/impactedresource.md:
```diff
@@ -0,0 +1,82 @@
+|Property|Type|Description|
+|:---|:---|:---|
+|addedDateTime|DateTimeOffset|The date and time when the [impactedResource](../resources/impactedresource.md) object was initially associated with a [recommendation](../resources/recommendation.md).|
+|additionalDetails|[keyValue](../resources/keyvalue.md) collection|Additional information unique to the [impactedResource](../resources/impactedresource.md) to help contextualize the [recommendation](../resources/recommendation.md).|
```
keyValue is a shared resource. The doc needs to be tweaked to indicate the different APIs that use it. I can push a comment to tweak the structure.

Q: Do you think it valuable to add a sample related to recommendations? You can add this update.

FaithOmbongi replied:
Sounds good. We'll update the doc once we have recommendations that use it.

FaithOmbongi commented on api-reference/beta/resources/impactedresource.md:
```diff
@@ -0,0 +1,82 @@
+|recommendationId|String|The unique identifier of the [recommendation](../resources/recommendation.md) that the resource is associated with.|
+|resourceType|String|Indicates the type of Azure AD resource. Examples include `user`, `application`.|
+|status|recommendationStatus|Indicates whether a resource needs to be addressed. The possible values are: `active`, `completedBySystem`, `completedByUser`, `dismissed`, `postponed`, `unknownFutureValue`. By default, a recommendation's **status** is set to `active` when the recommendation is first generated. **Status** is set to `completedBySystem` when our service detects that a resource which wa ...
+|subjectId|String|The related unique identifier, depending on the `resourceType`. For example, this property is set to the `applicationId` if the `resourceType` is an application.|
```
nit
```suggestion
|subjectId|String|The related unique identifier, depending on the **resourceType**. For example, this property is set to the `applicationId` if the **resourceType** is an `application`.|
```

FaithOmbongi replied:
@hafowler - bump

FaithOmbongi commented on api-reference/beta/resources/impactedresource.md:
```diff
@@ -0,0 +1,82 @@
+## Methods
+|Method|Return type|Description|
+|:---|:---|:---|
+|[Get impactedResource](../api/impactedresource-get.md)|[impactedResource](../resources/impactedresource.md)|Read the properties and relationships of an [impactedResource](../resources/impactedresource.md) object.|
```
Move the List impactedResource entry from recommendations.md to here.
```suggestion
|[List impactedResources](../api/recommendation-list-impactedresources.md)|[impactedResource](../resources/impactedresource.md) collection|Get the [impactedResource](../resources/impactedresource.md) objects for a [recommendation](recommendation.md).|
|[Get impactedResource](../api/impactedresource-get.md)|[impactedResource](../resources/impactedresource.md)|Read the properties and relationships of an [impactedResource](../resources/impactedresource.md) object.|
```

FaithOmbongi commented on api-reference/beta/api/recommendation-list-impactedresources.md:
```diff
@@ -0,0 +1,128 @@
+---
+title: "List impactedResources"
+description: "Get the impactedResource resources from the impactedResources navigation property."
```
```suggestion
description: "Get the impacted resources for a recommendation."
```

FaithOmbongi commented on api-reference/beta/api/recommendation-list-impactedresources.md:
```diff
@@ -0,0 +1,128 @@
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+Get the impactedResource resources from the impactedResources navigation property.
```
```suggestion
Get the [impactedResource](../resources/impactedresource.md) objects for a [recommendation](recommendation.md).
```

FaithOmbongi commented on api-reference/beta/api/recommendation-list-impactedresources.md:
````diff
@@ -0,0 +1,128 @@
+```
+
+## Optional query parameters
+This method supports some of the OData query parameters to help customize the response. For general information, see [OData query parameters](/graph/query-parameters).
````
👀

FaithOmbongi replied:
@hafowler - bump

FaithOmbongi commented on api-reference/beta/api/impactedresource-get.md:
````diff
@@ -0,0 +1,102 @@
+```
+
+## Optional query parameters
+This method supports some of the OData query parameters to help customize the response. For general information, see [OData query parameters](/graph/query-parameters).
````
.

FaithOmbongi replied:
@hafowler - Bump

hafowler replied:
@FaithOmbongi I don't understand the suggestion here?

FaithOmbongi commented on api-reference/beta/api/impactedresource-get.md:
````diff
@@ -0,0 +1,102 @@
+    ]
+}
+```
+
````
Add Example 2 using $select (at a minimum), $expand

FaithOmbongi replied:
@hafowler - Bump

FaithOmbongi commented on api-reference/beta/api/recommendation-list-impactedresources.md:
````diff
@@ -0,0 +1,128 @@
+    ]
+}
+```
+
````
Add Example 2 using $select (at a minimum), $expand, and $filter

FaithOmbongi replied:
@hafowler - bump

FaithOmbongi commented on api-reference/beta/api/impactedresource-postpone.md:
```diff
@@ -0,0 +1,114 @@
+## Request body
+In the request body, supply a JSON representation of the parameters.
+
+The following table shows the parameters that can be used with this action.
```
```suggestion
The following table shows the parameters that are required with this action.
```

FaithOmbongi commented on api-reference/beta/api/impactedresource-dismiss.md:
```diff
@@ -0,0 +1,114 @@
+## Request body
+In the request body, supply a JSON representation of the parameters.
+
+The following table shows the parameters that can be used with this action.
```
```suggestion
The following table shows the parameters that are required with this action.
```

FaithOmbongi commented on api-reference/beta/api/impactedresource-dismiss.md:
```diff
@@ -0,0 +1,114 @@
+
+|Parameter|Type|Description|
+|:---|:---|:---|
+|dismissReason|String|Audited reason for deeming an [impactedResource](../resources/impactedresource.md) inapplicable to you and its associated [recommendation](../resources/recommendation.md).|
```
```suggestion
|dismissReason|String|Audited reason for deeming an [impactedResource](../resources/impactedresource.md) and its associated [recommendation](../resources/recommendation.md) inapplicable to you.|
```

hafowler replied:
@FaithOmbongi, when you dismiss an impacted resource, you aren't dismissing the resource and the recommendation. I actually propose that I remove "its associated recommendation".

FaithOmbongi commented on api-reference/beta/toc.yml:
```diff
@@ -22032,6 +22032,39 @@ items:
               href: api/agreement-post-files.md
             - name: List localized files
               href: api/agreement-list-files.md
+    - name: Recommendations
```
What are your thoughts:
- Move this to "Identity and access" > "Directory management"
- Make "Recommendations" and "Impacted resources" peers in the TOC, instead of nesting.
```suggestion
    - name: Directory recommendations
```

FaithOmbongi commented on api-reference/beta/toc.yml:
```diff
@@ -22032,6 +22032,39 @@ items:
+        href: api/recommendation-postpone.md
+      - name: Reactivate
+        href: api/recommendation-reactivate.md
+      - name: impactedResource
+        items:
+        - name: impactedResource
```
```suggestion
      - name: Impacted resources
        items:
        - name: Impacted resource
```

FaithOmbongi commented on changelog/Microsoft.AAD.Reporting.json:
```diff
@@ -1,5 +1,127 @@
+      "Cloud": "Prod",
+      "Version": "beta",
+      "CreatedDateTime": "2023-01-09T22:16:18.601897Z",
+      "WorkloadArea": "Reports and audit",
```
```suggestion
      "WorkloadArea": "Reports",
```

FaithOmbongi commented on changelog/Microsoft.AAD.Reporting.json:
```diff
@@ -1,5 +1,127 @@
+          "ApiChange": "Resource",
+          "ChangedApiName": "recommendation",
+          "ChangeType": "Addition",
+          "Description": "Added the [recommendation](https://learn.microsoft.com/en-us/graph/api/resources/recommendation?view=graph-rest-beta) resource",
```
Add other objects for
- recommendationBase resource
- actionStep resource
- actionUrl resource
- impactedResource resource
- recommendation relationship to directory resource
- impactedResources relationship to directory resource


Delete

- I recommend deleting all the methods and enums. Because this is a new set of APIs, we can tweak the recommendation entry to "Added the recommendation resource and its associated methods." and "Added the impactedResource resource and its associated methods."

FaithOmbongi reviewed (changes requested):
Thank you @hafowler for your extensive work on this PR. I've left suggestions for changes inline (please verify technical accuracy) plus the following:

- Document the **recommendations** and **impactedResources** relationships in the [directory resource type](https://learn.microsoft.com/en-us/graph/api/resources/directory?view=graph-rest-beta#relationships).
- Document the recommendationBase entity.  This is the starting point for the recommendations set of APIs. The recommendation resource also inherits all its properties from recommendationBase.
    - PS: I see the Build for your schema PR has expired so you can't get doc stubs. I hope you did save them offline If not, please reach out to me to generate fresh ones for you.
- For property descriptions of the recommendation entity
     - Append "_Inherited from [recommendationBase](recommendationbase.md)._"
     - For properties that support $filter, indicate those properties and add the supported operators.
     - For relationships that support $expand, indicate those relationships.
     - Sample: [workflowTemplate](https://learn.microsoft.com/en-us/graph/api/resources/identitygovernance-workflowtemplate?view=graph-rest-beta)
- File name fix: recommendation-list.md =>directory-list-recommendation.md
- GLOBAL for API docs: Please list all RBAC roles that a user/app must have to call the respective APIs. I see at least 8 RBAC roles support read-only operations and 5 support Read/Write operations. For guidance on how to document, see this sample doc [List workflowTemplate](https://learn.microsoft.com/en-us/graph/api/identitygovernance-lifecycleworkflowscontainer-list-workflowtemplates?view=graph-rest-beta&tabs=http#permissions)
- GLOBAL: There's little documentation of supported query parameters. Please indicate within the property descriptions the ones that support $filter and $expand; For API docs, the "Optional query parameters" section should explicitly call out the supported query parameters ($select, $filter, $expand, $top, $skip etc). For guidance, see [workflows resource type](https://learn.microsoft.com/en-us/graph/api/resources/identitygovernance-workflow?view=graph-rest-beta) and the associated [Optional query parameters section of the List workflows API doc](https://learn.microsoft.com/en-us/graph/api/identitygovernance-lifecycleworkflowscontainer-list-workflows?view=graph-rest-beta&tabs=http#optional-query-parameters).
- GLOBAL for LIST and GET docs: Add an Example 2 that shows the use of $filter (for LIST only), $select., and $expand query parameters. At a minimum, Example 2 should show the use of $select.
   - Where you'll be tweaking the URLs, please re-run the requests in Graph Explorer.
   - Your Get recommendations doc shows some great examples but no. 2 and 3 are the correct syntaxes in the wrong doc. See more guidance in the inline comments.
- TOC Changes:
  - I suggest that this set of APIs be nested under "Identity and access" > "Directory management" sub-node. Then make the node name "Directory recommendations". What are your thoughts on this?
  - Also, make "Recommendations" and "Impacted resources" the same level (instead of nesting "Impacted resources")

- Changelog: I've left suggestions for missing objects in the changelog. Also delete some others. 
- We need a short API overview article to introduce this set of APIs. I'll work with you on this. This isn't non-blocking for this PR. We can publish it after we merge this PR.

hafowler commented:
@FaithOmbongi 
Here is what I have done:

	- Renamed from recommendation-list to directory-list-recommendations
	- Updated changelogs with your suggestions
	- Removed recommendations available list in recommendation.md
	- Removed impactedresources list method from recommendation.md and added to impactedresource.md
	- Applied nits
	- Updated TOC per your suggestion. Kept impacted resources as nested, since it isn't a relationship to the directory
	- Updated directory.md with recommendations relationship
	- Added changelog entries for other resources and recommendation relationship to directory
	- Added supported query parameters on properties and moved examples where appropriate
	- Added RBAC roles required

Here is what I didn't do:

	- Add impactedResources relationship to directory. A directory doesn't have impacted resources. A recommendation does.
	- Document recommendationBase entity. This is an internal concept that customers shouldn't be able to interface with. As such, I didn't add any "Inherited from… " lines

FaithOmbongi commented on api-reference/beta/api/directory-list-recommendation.md:
```diff
@@ -0,0 +1,298 @@
+doc_type: apiPageType
+---
+
+# List recommendationBases
```
nit
```suggestion
# List recommendations
```

FaithOmbongi commented on api-reference/beta/api/recommendation-get.md:
````diff
@@ -0,0 +1,136 @@
+    }
+  ]
+}
+```
````
Add example 2 using $select (required) and $expand

FaithOmbongi commented on api-reference/beta/resources/directory.md:
```diff
@@ -46,6 +46,7 @@ Inherits from [entity](entity.md).
 | featureRolloutPolicies             | [featureRolloutPolicy](featurerolloutpolicy.md) collection                               | Nullable.                                                                                                                 |
 | federationConfigurations           | [identityProviderBase](../resources/identityproviderbase.md) collection                  | Configure domain federation with organizations whose identity provider (IdP) supports either the SAML or WS-Fed protocol. |
 | onPremisesSynchronization          | [onPremisesDirectorySynchronization](../resources/onpremisesdirectorysynchronization.md) | A container for on-premises directory synchronization functionalities that are available for the organization.            |
+| recommendations | [recommendation](../resources/recommendation.md) collection | List of recommended improvements to improve tenant posture. |
```
I've seen your explanation as to why we aren't documenting impactedResources relationship. Does this mean the relationship was accidentally added (or was left over from the first round of schema updates)? If so, I recommend removing it from the schema - I don't think it'll be a breaking change because we won't document it anyway.

FaithOmbongi commented on api-reference/beta/resources/recommendation.md:
```diff
@@ -0,0 +1,109 @@
+
+## JSON representation
+The following is a JSON representation of the resource.
+<!-- {
+  "blockType": "resource",
+  "keyProperty": "id",
+  "@odata.type": "microsoft.graph.recommendation",
+  "openType": false
+}
```
@hafowler - bump

FaithOmbongi commented on api-reference/beta/resources/recommendationbase.md:
```diff
@@ -0,0 +1,90 @@
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+An abstract type that exposes the properties for configuring a custom lifecycle workflow. This resource is inherited by the [recommendation](../resources/recommendation.md) resource type.
```
```suggestion
An abstract type that exposes the properties for the [recommendation](../resources/recommendation.md) resource type.
```

FaithOmbongi commented on api-reference/beta/toc.yml:
```diff
@@ -20034,6 +20034,39 @@ items:
+            href: api/recommendation-postpone.md
+          - name: Reactivate
+            href: api/recommendation-reactivate.md
+          - name: Impacted resources
```
I understand your PoV because it's technically accurate. But we recommend 3 levels of nesting at most if possible - currently, "Directory recommendations" is at level 3. We should try to avoid level 4 if we can help it.

It's completely OK to not nest sometimes, like in this case.

Once we have an overview, both sub nodes will be preceded by the overview article and it'll improve the presentation.

FaithOmbongi reviewed (changes requested):
Hi @hafowler,

The major issue: I don't know how it came to be but somehow, you pulled changes from other branches into yours. Can you undo them before they mess up the state of your PR?

There were a few feedback items that were still pending so I've tagged you and bumped them up. 

One gap I forgot to point out: Please document the two new permissions in the permissions reference doc. You can copy what you did in the [old PR](https://github.com/microsoftgraph/microsoft-graph-docs/pull/14924/files#diff-ea9ba9ee5d35420a53064eefdbbe70054e6cd1a600d74f6ee1752a7de33c88ed).

We'll be good to publish afterward.

hafowler commented:
Closing this PR and [created a new one](https://github.com/microsoftgraph/microsoft-graph-docs/pull/20105).

FaithOmbongi commented:
The changes are live. Removing the changes requested label to avoid unactionable notifications.
